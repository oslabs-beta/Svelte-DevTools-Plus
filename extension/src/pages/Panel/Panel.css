html,
body {
  height: 100%;
  margin: 0;
  padding: 0;
}

body {
  --main-bg-color: #333b4a;
  --rewinder-height: 2.5rem;
  background-color: var(--main-bg-color);
}

.container {
  color: #eeeeee;
  border-radius: 1rem;
  box-shadow: 0px 0px 20px white;
}

.pane {
  display: flex;
  flex-direction: column;
  height: 100%;
  background-color: var(--main-bg-color);
  font-family: 'Source Sans Pro', sans-serif;
}

.nav-option {
  margin-right: 1rem;
  margin-left: 1rem;
  border-radius: 0.25rem;
}

.nav-option:hover {
  background-color: rgba(255, 255, 255, 0.1);
}

.nav-option:active {
  background-color: rgba(255, 255, 255, 0.4);
}

.split {
  display: flex;
  flex-direction: row;
}

.gutter {
  background-color: #eee;
  background-repeat: no-repeat;
  background-position: 50%;
}

.gutter.gutter-horizontal {
  width: 6px !important;
  background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAeCAYAAADkftS9AAAAIklEQVQoU2M4c+bMfxAGAgYYmwGrIIiDjrELjpo5aiZeMwF+yNnOs5KSvgAAAABJRU5ErkJggg==');
  cursor: col-resize;
}

.nav {
  display: flex;
  align-items: stretch;
}

.nav ul {
  display: flex;
  /* gap: 1rem; */
  list-style: none;
  margin: 0;
  padding: 0;
}

.nav a {
  color: inherit;
  text-decoration: none;
  display: flex;
  align-items: stretch;
  justify-content: center;
}

.nav li {
  color: rgb(28, 195, 221);
}

.nav-link {
  padding: 0.5rem;
  cursor: pointer;
  border-radius: 0.25rem;
}

.nav-link:active {
  background-color: rgb(55, 68, 114);
}
.nav-link:hover {
  background-color: rgb(114, 137, 218);
}

.page-content {
  height: calc(100% - var(--rewinder-height));
}

#treeWrapper,
.container,
.split {
  height: 100%;
}

.container {
  height: 100%;
}

.list-page-gap {
  height: 2rem;
}

header {
  height: 3.25rem;
  max-height: 3.25rem;
  border-bottom: #eeeeee solid 2px;
  display: flex;
  justify-content: center;
  align-items: center;
}

#no-data-error {
  text-align: center;
}

#left-pane {
  height: 100%;
}

.pane-content {
  overflow: overlay;
  height: calc(100% - var(--rewinder-height));
}

summary {
  cursor: pointer;
}

#last-update-message {
  position: absolute;
  color: yellow;
  margin-left: 1rem;
}
