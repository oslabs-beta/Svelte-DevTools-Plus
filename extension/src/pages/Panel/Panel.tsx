import React, { useEffect, useState } from 'react';
import './Panel.css';
import Split from 'react-split';
import ComponentInfo from './PanelComponents/ComponentInfo/ComponentInfo';
import Navbar from './PanelComponents/Navbar/Navbar';
import { Routes, Route, useNavigate } from 'react-router-dom';
import TreePage from './PanelPages/TreePage/TreePage';
import ListPage from './PanelPages/ListPage/ListPage';
import { Component } from './slices/highlightedComponentSlice';
import { useSelector } from 'react-redux';
import { selectCurrentSnapshot } from './slices/currentSnapshotSlice';
import { useDispatch } from 'react-redux';
import { TreeHistory, selectTreeHistory } from './slices/treeHistorySlice';
import Rewinder from './PanelComponents/Rewinder/Rewinder';
import { TimestampState, selectTimestamps } from './slices/timestampSlice';

export interface ComponentPageProps {
  rootComponentData: Component;
}

// Let the rest of the extension know that the panel is closed
// so it won't try and send messages to it
window.addEventListener('beforeunload', function () {
  chrome.runtime.sendMessage({ message: 'handleClosedPanel' });
});

function Panel() {
  const currentSnapshot = useSelector(selectCurrentSnapshot);
  const treeHistory: TreeHistory = useSelector(selectTreeHistory);
  const timestamps: TimestampState = useSelector(selectTimestamps);
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const [unableToGetComponentData, setUnableToGetComponentData] =
    useState(false);

  // Navigate to the root directory on page load
  useEffect(() => {
    navigate('/');
  }, []);

  console.log(timestamps);

  useEffect(() => {
    async function setUpPanel() {
      try {
        const [tab] = await chrome.tabs.query({
          active: true,
          lastFocusedWindow: true,
        });
        if (tab && tab.id !== undefined) {
          chrome.tabs.sendMessage(tab.id, { message: 'getRootComponent' });
        }
      } catch (err) {
        console.log(err);
      }
    }

    // I only want to add a listener once, so qit goes in the onMount useEffect
    // Listens for response from ContentScriptIsolated. This is where we
    // get the current tab's root component, and process updates
    function messageListener(message: any) {
      dispatch({
        type: 'timestamps/addNewTimestamp',
        payload: {
          timestamp: console.time(), // TODO: Should be timeEnd(). Fix this
        },
      });
      if (message.type === 'updateRootComponent') {
        const rootComponent = message.rootComponent;
        if (rootComponent) {
          createAndSaveNewSnapshot(rootComponent);
        }
      } else if (message.type === 'returnRootComponent') {
        const rootComponent = message.rootComponent;
        if (rootComponent) {
          createAndSaveNewSnapshot(rootComponent);
        } else {
          setUnableToGetComponentData(true);
        }
        // For use after rewinding
      } else if (message.type === 'returnTempRoot') {
        const tempRoot = message.rootComponent;
        // set the tempRoot as the current snapshot without saving it
        dispatch({
          type: 'currentSnapshot/setCurrentSnapshot',
          payload: {
            rootComponent: tempRoot,
          },
        });
      }
    }
    chrome.runtime.onMessage.addListener(messageListener);
    setUpPanel();
    return () => {
      chrome.runtime.onMessage.removeListener(messageListener);
    };
  }, []);

  function createAndSaveNewSnapshot(newRootComponent: Component) {
    dispatch({
      type: 'currentSnapshot/setCurrentSnapshot',
      payload: {
        rootComponent: newRootComponent,
      },
    });
    dispatch({
      type: 'treeHistory/addNewSnapshot',
      payload: {
        newSnapshot: newRootComponent,
      },
    });
  }

  /*
    Takes the snapshot in treeHistory[snapshotIndex] and injects state with all
    of its component's state. Then the content script returns a "temp root component"
    which gets displayed to the screen without saving a new snapshot
  */
  async function changeSnapshot(snapshotIndex: number) {
    try {
      const [tab] = await chrome.tabs.query({
        active: true,
        lastFocusedWindow: true,
      });
      chrome.tabs.sendMessage(tab.id!, {
        message: 'injectSnapshot',
        snapshot: treeHistory.treeHistory[snapshotIndex],
      });
    } catch (err) {
      console.log('Error getting tab: ', err);
    }
  }

  function clearSnapshotHistory() {
    dispatch({
      type: 'treeHistory/deleteAllSnapshots',
    });
  }

  return (
    <div className="container" data-testid="panel">
      <div className="page-content">
        <Split className="split">
          <div className="pane">
            <header>
              <Navbar />
            </header>
            {unableToGetComponentData ? (
              <h1 id="no-data-error" data-testid="no-data-error">
                Unable to get component data
              </h1>
            ) : (
              <Routes>
                <Route
                  path="/"
                  element={
                    <ListPage
                      rootComponentData={currentSnapshot.rootComponent!}
                    />
                  }
                />
                <Route
                  path="/tree"
                  element={
                    <TreePage
                      rootComponentData={currentSnapshot.rootComponent!}
                    />
                  }
                />
              </Routes>
            )}
          </div>
          <div>
            <ComponentInfo />
          </div>
        </Split>
      </div>
      <Rewinder
        changeSnapshot={changeSnapshot}
        numberOfSnapshots={treeHistory.treeHistory.length}
        clearSnapshotHistory={clearSnapshotHistory}
      />
    </div>
  );
}

export default Panel;
