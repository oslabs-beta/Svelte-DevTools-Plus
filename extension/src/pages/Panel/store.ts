import { configureStore } from '@reduxjs/toolkit';
import highlightedComponentReducer from './slices/highlightedComponentSlice';
import currentSnapshotReducer from './slices/currentSnapshotSlice';
import treeHistoryReducer from './slices/treeHistorySlice';

export const store = configureStore({
  reducer: {
    highlightedComponent: highlightedComponentReducer,
    currentSnapshot: currentSnapshotReducer,
    treeHistory: treeHistoryReducer,
  },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
