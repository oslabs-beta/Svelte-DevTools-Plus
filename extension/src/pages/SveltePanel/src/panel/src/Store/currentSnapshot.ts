import { currentSnapshot } from "./store";
import { Component } from '../types';

export function setCurrentSnapshot(rootComponent: Component) {
  currentSnapshot.update(state => {
    return { ...state, rootComponent };
  });
}