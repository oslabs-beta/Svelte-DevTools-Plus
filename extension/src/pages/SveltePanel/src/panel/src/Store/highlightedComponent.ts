import { highlightedComponent } from './store';
import type { Component } from '../types';

export function setHighlightedComponent(payload: Component) {
  highlightedComponent.set(payload);
}

export function updateHighlightedComponent(payload: Component) {
  highlightedComponent.update((current) => ({
    ...current,
    detail: payload.detail,
    tagName: payload.tagName,
  }));
}
