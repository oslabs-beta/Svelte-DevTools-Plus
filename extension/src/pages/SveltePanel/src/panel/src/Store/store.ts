import { writable } from 'svelte/store';
import type {
  Component,
  Snapshot,
  TimedEventsState,
  TreeHistory,
} from '../types';

// Define initial states
const highlightedComponentInitial: Component = {
  tagName: '',
  detail: [],
  children: [],
  id: -1,
};

const currentSnapshotInitial = {
  rootComponent: null,
};

const treeHistoryInitial = {
  treeHistory: [],
};

const timedEventsInitial = {
  newTimeStart: -1,
  eventTimes: [],
};

// Create stores
export const highlightedComponent = writable<Component>(
  highlightedComponentInitial
);
export const currentSnapshot = writable<Snapshot>(currentSnapshotInitial);
export const treeHistory = writable<TreeHistory>(treeHistoryInitial);
export const timedEvents = writable<TimedEventsState>(timedEventsInitial);
