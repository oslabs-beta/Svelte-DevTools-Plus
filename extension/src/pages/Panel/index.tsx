import React from 'react';
import { createRoot } from 'react-dom/client';
import { Provider } from 'react-redux';
import Panel from './Panel';
import './index.css';
import { store } from '../../../Store/store';
import { BrowserRouter } from 'react-router-dom';

const container = document.getElementById('app-container');
const root = createRoot(container!);
root.render(
  <Provider store={store}>
    <BrowserRouter>
      <Panel />
    </BrowserRouter>
  </Provider>
);
