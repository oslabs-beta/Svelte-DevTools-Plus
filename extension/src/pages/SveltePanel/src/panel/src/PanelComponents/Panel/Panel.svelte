<script lang="ts">
  import { onDestroy, onMount } from 'svelte';
  import Navbar from '../Navbar/Navbar.svelte';
  import ComponentInfo from '../ComponentInfo/ComponentInfo.svelte';
  import TreePage from '../../PanelPages/TreePage/TreePage.svelte';
  import ListPage from '../../PanelPages/ListPage/ListPage.svelte';
  import Rewinder from '../Rewinder/Rewinder.svelte';
  import {
    currentSnapshot,
    treeHistory,
  } from '../../Store/store';
  import sendMessageToChrome from '../../../../../../../messenger';
  import { addNewSnapshot, deleteAllSnapshots } from '../../Store/treeHistory';
  import { setCurrentSnapshot } from '../../Store/currentSnapshot';
    import { addNewEvent } from '../../Store/timedEvents';
    import { get } from 'svelte/store';

  let unableToGetComponentData = false;
  let lastUpdateMessage = '';

  window.addEventListener('beforeunload', () => {
    sendMessageToChrome('handleClosedPanel');
  });

  window.addEventListener('message', (event: MessageEvent) => {
    const { type, rootComponent } = event.data;
    if (type === 'updateRootComponent' && rootComponent) {
      setCurrentSnapshot(rootComponent);
      addNewSnapshot(rootComponent);
    } else if (type === 'returnRootComponent') {
      if (!rootComponent) {
        unableToGetComponentData = true;
      } else {
        setCurrentSnapshot(rootComponent);
      }
    } else if (type === 'returnProfilingData') {
      lastUpdateMessage = `Last update took ${event.data.time}ms`;
    }
  });

  async function changeSnapshot(snapshotIndex: number) {
    try {
      console.log('chrome', chrome);
      const [tab] = await chrome.tabs.query({
        active: true,
        lastFocusedWindow: true,
      });
      addNewEvent({
          type: 'sendMessage',
          data: performance.now(),
        })
      sendMessageToChrome('injectSnapshot', {
        snapshot: get(treeHistory).treeHistory[snapshotIndex],
        tab,
      });
    } catch (err) {
      console.log('Error getting tab: ', err);
    }
  }

  function clearSnapshotHistory() {
    deleteAllSnapshots();
  }


  onDestroy(() => {
    window.removeEventListener('beforeunload', () => {});
    window.removeEventListener('message', () => {});
  });
</script>

<svelte:head>
  <link rel="stylesheet" href="./Panel.css" />
</svelte:head>

<div class="container" data-testid="panel">
  <div class="split">
    <div class="pane">
      <header>
        <Navbar />
      </header>
      {#if unableToGetComponentData}
        <h1 id="no-data-error" data-testid="no-data-error">
          Unable to get component data
        </h1>
      {:else}
        <nav>
          <div id="left-pane">
            <a href="/">
              <ListPage rootComponentData={get(currentSnapshot)} />
            </a>
            <a href="/tree">
              <TreePage rootComponentData={get(currentSnapshot)} />
            </a>
          </div>
        </nav>
        <p id="last-update-message">{lastUpdateMessage}</p>
      {/if}
    </div>
    <div>
      <ComponentInfo />
    </div>
  </div>
  <Rewinder
    {changeSnapshot}
    numberOfSnapshots={get(treeHistory).treeHistory.length}
    {clearSnapshotHistory}
  />
</div>
