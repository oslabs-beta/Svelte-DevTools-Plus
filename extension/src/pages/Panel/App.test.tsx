/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { jest } from '@jest/globals';
import { act } from 'react-dom/test-utils';
import userEvent from '@testing-library/user-event';
import Panel from './Panel';
import { createTestStore } from '../../../Store/store';

jest.mock('chrome');

async function customRender(ui: any, store: any) {
  await act(async () =>
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Panel />
        </BrowserRouter>
      </Provider>
    )
  );
}

describe('Panel tests', function () {
  let store: any;
  beforeEach(async () => {
    store = createTestStore();
    await customRender(<Panel />, store);
  });

  afterEach(() => {
    chrome.clearListeners();
    chrome.resetMockData();
    jest.clearAllMocks();
  });

  it('Successfully loads component data', () => {
    const app = screen.getByText('App');
    expect(app).toBeInTheDocument();
  });

  it('Matches the snapshot', () => {
    const element = screen.getByTestId('panel');
    expect(element).toMatchSnapshot();
  });

  it('Expands and collapses the child components', async () => {
    // Expand everything
    const appButton = screen.getByTestId('expand-button-App');
    await userEvent.click(appButton);
    // App's children are now expanded
    let boardButton: HTMLElement | null = screen.getByTestId(
      'expand-button-Board'
    );
    await userEvent.click(boardButton);
    // Board's children are now expanded
    let rowButtons = screen.getAllByTestId('expand-button-Row');
    expect(rowButtons.length).toBe(3);
    for (let row = 0; row < rowButtons.length; row++) {
      const rowButton = rowButtons[row];
      await userEvent.click(rowButton);
      const boxButtons = screen.getAllByTestId('component-leaf-Box');
      expect(boxButtons.length).toBe(3 * (row + 1));
    }
    // All Row children are now expanded
    // Collapse everything
    for (let row = 0; row < rowButtons.length; row++) {
      const rowButton = rowButtons[row];
      await userEvent.click(rowButton);
      const boxButtons = screen.queryAllByTestId('component-leaf-Box');
      expect(boxButtons.length).toBe(3 * (rowButtons.length - row - 1));
    }
    // All Row children are now collapsed
    await userEvent.click(boardButton);
    // Board's children are now collapsed
    rowButtons = screen.queryAllByTestId('expand-button-Row');
    expect(rowButtons.length).toBe(0);
    await userEvent.click(appButton);
    // App's children are now collapsed
    boardButton = screen.queryByTestId('expand-button-Board');
    expect(boardButton).toBeNull();
  });

  it('Properly displays component info', async () => {
    // Open App component
    const appExpand = screen.getByTestId('expand-button-App');
    await userEvent.click(appExpand);
    // Click on Board component
    const boardButton = screen.getByTestId('component-button-Board');
    await userEvent.click(boardButton);
    // Check Board's turn state
    const turnStateButton = screen.getByTestId('state-value-turn');
    expect(turnStateButton.querySelector('p')?.innerHTML).toBe('X');
  });

  async function changeBoardTurnState(value: string) {
    const turnModifier = screen.getByTestId('modifier-turn');
    await userEvent.click(turnModifier);
    await userEvent.type(turnModifier, value);
    await userEvent.keyboard('{Enter}');
  }

  it('Changes turn state to "Q"', async () => {
    // Click on Board component
    const appExpand = screen.getByTestId('expand-button-App');
    await userEvent.click(appExpand);
    const boardButton = screen.getByTestId('component-button-Board');
    await userEvent.click(boardButton);
    const value = 'Q';
    await changeBoardTurnState(value);
    // Check if its state has been changed
    const turnStateButton = screen.getByTestId('state-value-turn');
    expect(turnStateButton.querySelector('p')?.innerHTML).toBe(value);
  });

  it('Rewinds and reverts state', async () => {
    // Click on Board component
    const appExpand = screen.getByTestId('expand-button-App');
    await userEvent.click(appExpand);
    const boardButton = screen.getByTestId('component-button-Board');
    await userEvent.click(boardButton);
    // Change its state to create new snapshots
    await changeBoardTurnState('1');
    await changeBoardTurnState('2');
    await changeBoardTurnState('3');
    // Check if its state has changed
    let turnStateButton = screen.getByTestId('state-value-turn');
    expect(turnStateButton.querySelector('p')?.innerHTML).toBe('3');
    let rewindButton = screen.getByTestId('rewind-button');
    await userEvent.click(rewindButton);
    // Check if rewind was successful
    turnStateButton = screen.getByTestId('state-value-turn');
    expect(turnStateButton.querySelector('p')?.innerHTML).toBe('2');
    const forwardButton = screen.getByTestId('revert-button');
    await userEvent.click(forwardButton);
    // Check if revert was successful
    turnStateButton = screen.getByTestId('state-value-turn');
    expect(turnStateButton.querySelector('p')?.innerHTML).toBe('3');
    // Clicks the mock slider, which sets the snapshot to snapshot #2
    const rewinderSlider = screen.getByTestId('rewinder-slider');
    await userEvent.click(rewinderSlider);
    turnStateButton = screen.getByTestId('state-value-turn');
    expect(turnStateButton.querySelector('p')?.innerHTML).toBe('1');
    // Delete all snapshots
    const clearButton = screen.getByTestId('clear-button');
    await userEvent.click(clearButton);
    // Clicking the Slider shouldn't change anything
    await userEvent.click(rewinderSlider);
    expect(turnStateButton.querySelector('p')?.innerHTML).toBe('1');
    // Going forward shouldn't do anything
    await userEvent.click(forwardButton);
    expect(turnStateButton.querySelector('p')?.innerHTML).toBe('1');
    // Going back shouldn't do anything
    await userEvent.click(rewindButton);
    expect(turnStateButton.querySelector('p')?.innerHTML).toBe('1');
  });

  // Jest does not support svgdom so this is all the testing we can
  // do with TreePage
  it('Navigates to TreePage', async () => {
    const treeButton = screen.getByTestId('tree-link');
    await userEvent.click(treeButton);
    const treePage = screen.getByTestId('tree-page');
    expect(treePage).toBeInTheDocument();

    // This tells our chrome mock to return empty data on the next test
    // 'Displays an error message if it can not get component data' must
    // come after this function call!
    chrome.sendEmptyDataOnNextRequest();
  });

  it('Displays an error message if it can not get component data', async () => {
    const noDataError = screen.getByTestId('no-data-error');
    expect(noDataError.innerHTML).toBe('Unable to get component data');
  });
});
