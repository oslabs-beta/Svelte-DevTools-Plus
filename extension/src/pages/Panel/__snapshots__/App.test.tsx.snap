// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel tests Matches the snapshot 1`] = `
<div
  class="container"
  data-testid="panel"
>
  <div
    class="page-content"
  >
    <div
      class="split"
    >
      <div
        class="pane"
        style="width: calc(50% - 5px);"
      >
        <header>
          <nav
            class="nav"
          >
            <div
              class="nav-option"
            >
              <a
                aria-current="page"
                class="nav-link active"
                href="/"
                style="color: rgb(28, 195, 221); background: rgba(50, 50, 50, 0.25);"
              >
                List
              </a>
            </div>
            <div
              class="nav-option"
            >
              <a
                class="nav-link"
                data-testid="tree-link"
                href="/tree"
                style="color: rgb(255, 255, 255); background: rgba(0, 0, 0, 0);"
              >
                Tree
              </a>
            </div>
          </nav>
        </header>
        <div
          class="pane-content"
        >
          <div
            data-testid="root-container"
            id="root-container"
          >
            <div
              tabindex="0"
            >
              <div>
                <div
                  class="tree-component"
                >
                  <button
                    class="expand-button-container"
                    data-testid="expand-button-App"
                  >
                    <img
                      class="expand-button"
                      src="test-file-stub"
                    />
                  </button>
                  <button
                    class="tree-component-bar"
                    data-testid="component-button-App"
                  >
                    &lt;
                    <span
                      class="component-name"
                    >
                      App
                    </span>
                    /&gt;
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="gutter gutter-horizontal"
        style="width: 10px;"
      />
      <div
        style="width: calc(50% - 5px);"
      >
        <div
          class="pane"
        >
          <header>
            <h2 />
          </header>
          <div
            class="pane-content"
          >
            <h3>
              State
            </h3>
            <h3>
              Props
            </h3>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    id="rewinder"
  >
    <button
      data-testid="clear-button"
      id="clear-button"
    >
      Clear
    </button>
    <div
      id="slider-container"
    >
      <div
        data-testid="rewinder-slider"
      />
    </div>
    <button
      data-testid="rewind-button"
      id="back-button"
    >
      &lt;&lt;
    </button>
    <button
      data-testid="revert-button"
      id="forward-button"
    >
      &gt;&gt;
    </button>
  </div>
</div>
`;
